### ISSUE TITLE
Possible simplification to NATDIVSUCAUX

### ISSUE THREAD

**Issue by oisdk** [agda/agda#4991]
Currently the built-in `div-helper` function needs to have the following definition:

```agda
div-helper : ℕ → ℕ → ℕ → ℕ → ℕ
div-helper k m  zero    j      = k
div-helper k m (suc n)  zero   = div-helper (suc k) m n m
div-helper k m (suc n) (suc j) = div-helper k m n j
```

But the `k` accumulating parameter is unnecessary, as we could apply the `suc` to the recursive call:

```agda
div-helper′ : (m n j : ℕ) → ℕ
div-helper′ m  zero    j      = zero
div-helper′ m (suc n)  zero   = suc (div-helper′ m n m)
div-helper′ m (suc n) (suc j) = div-helper′ m n j
```

I have found this definition much easier to write proofs about.

**Comment by gallais** [comment]
I suspect this would introduce a space leak

**Comment by UlfNorell** [comment]
This function only ever run on open terms at type-checking time, so I wouldn't worry about performance. Also I actually think that the new version is better in that regard, since `suc` is strict but `div-helper` is not.

**Comment by andreasabel** [comment]
> I have found this definition much easier to write proofs about.

Hardly surprising.  ;-)  Functions using accumulators are in general tedious to verify.

**Comment by andreasabel** [comment]
Clearly not a refactoring, rather a small change of the Agda2 language.

**Comment by oisdk** [comment]
w.r.t the space leak, doesn't the new version have the same behaviour as the standard definition of `_+_`?

```agda
zero  + m = m
suc n + m = suc (n + m)
```

I had thought that both were tail recursion modulo cons, and therefore didn't leak space. (whereas the version with the accumulator actually does leak space because of laziness)

**Comment by MatthewDaggitt** [comment]
Assuming the new definition doesn't have any major problems, this change gets a big thumbs from me. [`Data.Nat.DivMod.Core`](https://github.com/agda/agda-stdlib/blob/master/src/Data/Nat/DivMod/Core.agda) was possibly some of the most painful Agda I've ever written.

**Comment by andreasabel** [comment]
@UlfNorell wrote:
> Also I actually think that the new version is better in that regard, since `suc` is strict but `div-helper` is not.

@oisdk wrote:
> (whereas the version with the accumulator actually does leak space because of laziness)

I am not following the arguments on space leaks.
Isn't the weak head normal form of the old `divhelper 0 m n j` always a numeral `suc^l 0`?
(The same holds for the new `divhelper' m n j` if `suc` is strict, otherwise not, so I guess that is why @gallais brought up the space leak.)

**Comment by UlfNorell** [comment]
The space leak occurs as you run the function, building up unevaluated calls to `suc` in the accumulator argument. The new version builds up `suc`-frames on the evaluation stack instead, which is slightly more efficient.

But, I stand by my initial comment that worrying about performance of this function is silly, since it won't run in the cases where you care about performance. Making it easier to prove things about it is a major win though.

**Comment by gallais** [comment]
Right, I guess my mistake was to believe that we were going to run that Haskell
code in the backend when we actually use some magic to efficiently compute the
result of `div-helper` on closed terms.

**Comment by andreasabel** [comment]
PR welcome!
